File: mobile/truebeauty/lib/services/api_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:truebeauty/utilities/api_url.dart';

class ApiService {
  final String baseUrl = ApiUrls.baseUrl;

  ApiService();

  Future<dynamic> get(String endpoint) async {
    final response = await http.get(Uri.parse('$baseUrl$endpoint'));

    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else {
      throw Exception('GET request failed: ${response.statusCode}');
    }
  }

  Future<dynamic> post(String endpoint, Map<String, dynamic> data) async {
    final response = await http.post(
      Uri.parse('$baseUrl$endpoint'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode(data),
    );

    if (response.statusCode == 200 || response.statusCode == 201) {
      return jsonDecode(response.body);
    } else {
      throw Exception('POST request failed: ${response.statusCode}');
    }
  }

  Future<dynamic> put(String endpoint, Map<String, dynamic> data) async {
    final response = await http.put(
      Uri.parse('$baseUrl$endpoint'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode(data),
    );

    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else {
      throw Exception('PUT request failed: ${response.statusCode}');
    }
  }

  Future<void> delete(String endpoint) async {
    final response = await http.delete(Uri.parse('$baseUrl$endpoint'));

    if (response.statusCode != 200) {
      throw Exception('DELETE request failed: ${response.statusCode}');
    }
  }
}

File: client/src/app/services/api-service.ts
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ApiUrls } from '../utilities/api-url';

@Injectable({
  providedIn: 'root',
})
export class ApiService {
  apiUrl = ApiUrls.baseUrl;
  http = inject(HttpClient);

  request(
    endpoint: string,
    method: string,
    payload: any = null
  ): Observable<any> {
    const url = `${this.apiUrl}${endpoint}`;
    const headers = this.createHeaders();

    return this.makeHttpRequest(url, method, payload, headers);
  }

  downloadFile(endpoint: string, payload: any = null): Observable<Blob> {
    const url = `${this.apiUrl}${endpoint}`;
    const headers = this.createHeaders();

    return this.http.post(url, payload, {
      headers,
      responseType: 'blob', // important for binary files
      withCredentials: true,
    });
  }

  private createHeaders(): HttpHeaders {
    const accessToken = localStorage.getItem('cims-accessToken');

    let headers = new HttpHeaders({
      'Content-Type': 'application/json',
    });

    if (accessToken) {
      headers = headers.set('Authorization', `Bearer ${accessToken}`);
    }

    return headers;
  }

  private makeHttpRequest(
    url: string,
    method: string,
    payload: any,
    headers: HttpHeaders
  ): Observable<any> {
    const options = { headers, withCredentials: true }; // Add withCredentials

    switch (method.toUpperCase()) {
      case 'POST':
        return this.http.post(url, payload, options);
      case 'GET':
        return this.http.get(url, options);
      case 'PUT':
        return this.http.put(url, payload, options);
      case 'DELETE':
        return this.http.delete(url, options);
      default:
        throw new Error(`Unsupported request method: ${method}`);
    }
  }
}
